crates/sui-types/src/deny_list_v2.rs

```diff
@@ -74,43 +78,68 @@ impl MoveTypeTagTrait for AddressKey {
     }
 }
 
+pub fn check_coin_deny_list_v2_during_signing(
+    address: SuiAddress,
+    input_objects: &CheckedInputObjects,
+    receiving_objects: &ReceivingObjects,
+    object_store: &dyn ObjectStore,
+) -> UserInputResult {
+    let coin_types = input_object_coin_types_for_denylist_check(input_objects, receiving_objects);
+    for coin_type in coin_types {
+        // TODO: Check global pause flag.
+        let Some(deny_list) = get_per_type_coin_deny_list_v2(&coin_type, object_store) else {
+            return Ok(());
+        };
+        if check_address_denied_by_coin(&deny_list, address, object_store, None) {
+            return Err(UserInputError::AddressDeniedForCoin { address, coin_type });
+        }
+    }
+    Ok(())
+}
+
 pub fn get_per_type_coin_deny_list_v2(
-    coin_type: String,
+    coin_type: &String,
     object_store: &dyn ObjectStore,
 ) -> Option<Config> {
     let deny_list_root =
         get_deny_list_root_object(object_store).expect("Deny list root object not found");
-    let config: Config = get_dynamic_field_from_store(
-        object_store,
-        deny_list_root.id(),
-        &ConfigKey {
+    let config_key = DOFWrapper {
+        name: ConfigKey {
             per_type_index: DENY_LIST_COIN_TYPE_INDEX,
             per_type_key: coin_type.as_bytes().to_vec(),
         },
-    )
-    .ok()?;
+    };
+    let config: Config =
+        get_dynamic_field_from_store(object_store, deny_list_root.id(), &config_key).ok()?;
     Some(config)
 }
 
 pub fn check_address_denied_by_coin(
     coin_deny_config: &Config,
     address: SuiAddress,
     object_store: &dyn ObjectStore,
-    cur_epoch: EpochId,
+    cur_epoch: Option<EpochId>,
 ) -> bool {
     let address_key = AddressKey(address);
     read_config_setting(object_store, coin_deny_config, address_key, cur_epoch).unwrap_or(false)
 }
 
+/// Read the setting value from the config object in the object store for the given setting name.
+/// If the setting is not found, return None.
+/// If the setting is found, and if cur_epoch is not specified, it means we always want the latest value,
+/// so return the newer value.
+/// If the setting is found, and if cur_epoch is specified, it means we want the value prior to cur_epoch,
+/// so return the older value if the newer_value_epoch is equal to or newer than cur_epoch,
+/// otherwise return the newer value.
 fn read_config_setting<K, V>(
     object_store: &dyn ObjectStore,
     config: &Config,
     setting_name: K,
-    cur_epoch: EpochId,
+    cur_epoch: Option<EpochId>,
```

Comment:
I ended up undoing the Option bit, because you _should_ always have an epoch ID to query with? 
We can add it back if you like